src/liballoc/collections/linked_list.rs

```diff
@@ -143,14 +146,18 @@ impl<T> LinkedList<T> {
     /// Adds the given node to the front of the list.
     #[inline]
     fn push_front_node(&mut self, mut node: Box<Node<T>>) {
+        // This method takes care not to create mutable references, to maintain
+        // validity of aliasing pointers into existing nodes.
         unsafe {
             node.next = self.head;
             node.prev = None;
             let node = Some(Box::into_raw_non_null(node));
 
             match self.head {
                 None => self.tail = node,
-                Some(mut head) => head.as_mut().prev = node,
+                // Not creating new mutable (unique!) references to not invalidate
+                // references we handed out.
+                Some(head) => (*head.as_ptr()).prev = node,
```

Comment:
I'm not sure I even understand what is going on here...

Previously, `as_mut` returned a &mut Node<T>`, which we then modified the `prev` field of. That much makes sense.

Now, we get a `*mut Node<T>`, dereference it to get a `Node<T>` then set the `prev` field of.

I've been under the impression that setting the field of a struct implicitly created a mutable reference — where am I wrong? If I'm not wrong, why aren't we creating a second mutable reference here?

Reply:
I believe a mutable borrow is established only if you create an &mut that isn't *immediately* casted to a *mut. It's very janky immediate-action rules, with similar logic to why you can take a reference to the output of `arr[x]` and it's not a reference to a temporary.

Regardless for a lot of these the issue isn't a mutable borrow, but rather a mutable borrow that overlaps with `node.elem`. By writing the code this way we never claim unique ownership of the whole type, just that we can mutate the next/prev links.

Reply:
> I believe a mutable borrow is established only if you create an &mut that isn't immediately explicitly cast to a *mut.

No, that would be https://github.com/rust-lang/rfcs/pull/2582 which is not yet in.

> I've been under the impression that setting the field of a struct implicitly created a mutable reference

Writing to a field has all the same effects (in terms of invalidating other references) as creating a mutable reference to it, yes.

The key point is that *we only modify the `prev` field*, and the aliasing reference we are worried about points to the data field (`element` or whatever it is called).  I will expand the comments to clarify this.

> By writing the code this way we never claim unique ownership of the whole type, just that we can mutate the next/prev links.

s/type/node/. But yes.

Reply:
@shepmaster I tried to improve the comments explaining why these changes are needed; does this make more sense now?

Reply:
> Writing to a field has all the same effects (in terms of invalidating other references) as creating a mutable reference to it.

Can you remove the pronoun here? Do you mean:

- Writing to a struct's field has the same effects as creating a mutable reference to
  - the struct
  - the field

Reply:
I mean "to the field".

It's really the other way around -- creating a mutable reference has the effects of a write to the memory that is "covered" by the reference (`size_of::<T>` bytes starting at where the pointer points).